### ISSUE TITLE
Consistency regarding the total upload/download traffic on System Status page

### ISSUE THREAD

**Issue by sgrtye** [alireza0/x-ui#789]
**Describe the bug**
The total upload/download traffic on the System Status page claims to represent the total network traffic since the system startup. However, it represents the total traffic since instance x-ui has been running. If the x-ui is started after the system startup or is updated to another version, the total traffic is reset to zero. Meanwhile, the OS Operation Hours remain accurate, thus, creating inconsistency between the claimed total traffic and represented total traffic.

**To Reproduce**
1. Having a running instance of x-ui in docker.
2. Update the stack/container.
3. See the 'total traffic since system startup' is reset to zero.

**Expected behavior**
Having consistency between claimed and presented traffic data.

My proposed solution would be to change the description to total traffic since X-ray startup, and whenever the X-ray core is restarted, reset the traffic indicator.

**Screenshots**
<img width="839" alt="1703019843149" src="https://github.com/alireza0/x-ui/assets/55668018/f8921b2a-9637-46e6-9d58-198742ac5a06">

**Desktop (please complete the following information):**
 - OS: Ubuntu 22.04
 - Browser: Chrome
 - Version: Latest

**Smartphone (please complete the following information):**
 - Device: iPhone
 - OS: Latest
 - Browser: Chrome
 - Version: Latest

**Additional context**
I also noticed in the bug report template, a few colons are missing for the Browser/Version part.

**Comment by sgrtye** [comment]
> my server is up for almost 7 days and in the meantime updated two times, first to X-UI 1.6.3 and then 1.6.4. I have not see any problem regarding the issue you described above. my traffic usage works just fine. ![135E3DAA-EE50-4752-B396-64B2CC398D6E](https://private-user-images.githubusercontent.com/115543613/291716861-e9245844-b83c-4a22-bce1-a9989085baef.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwMjEzNTYsIm5iZiI6MTcwMzAyMTA1NiwicGF0aCI6Ii8xMTU1NDM2MTMvMjkxNzE2ODYxLWU5MjQ1ODQ0LWI4M2MtNGEyMi1iY2UxLWE5OTg5MDg1YmFlZi5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTlUMjEyNDE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWUzMTAwYTI3NmUxMTI2YTE1NWNlNzI4YmIxNDYxYTQ1YTZlNzBmZWY3OTY1NTMzYjNkZWYwNGMxNmIxZjE2MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.2KQjEZFCxOPlFYZAMF5atktK2p1K13sCidkeIAcmtck)

Thanks for your information, are you running it in a docker environment or as a standalone service? My problem was caused by updating the docker image.

**Comment by alireza0** [comment]
Dear @sgrtye ,

Please make sure that your problem is a bug. You can ask an expert for your problems.

When you are using docker, you should know that it is ephemeral virtualization. That means when you restart the container, another container will be created! Therefore, reseting total traffic is expected always.

In each restart panel (including restart after update), it will restart the whole docker. Thus, this behaviour is normal.

Main page contains all the data that application will get from the system. In Inbound page, information is related to the data which is received from database. No consistency/equality could be expected here.

**Comment by sgrtye** [comment]
Dear @alireza0,

I appreciate your prompt response and clarification regarding the behavior of the total upload/download traffic on the System Status page, particularly in Docker environments. While I acknowledge the ephemeral nature of Docker containers and the inherent reset of total traffic upon container restarts, I would like to share my perspective on the matter.

In my understanding, Docker serves as a tool to eliminate constraints and streamline the deployment of software, aiming for consistent behavior whether installed natively or through Docker. When employing Docker with the latest tag and updating the image, it is analogous to updating a standalone application. Consequently, I believe that the behavior concerning total traffic should remain consistent in both scenarios.

The problem cannot be fully explained by the provided rationale regarding the creation of a new container equating to a new OS. I observe a notable incongruity in the interpretation of system uptime—specifically, the System Uptime OS reflects the uptime of the physical server rather than the container. If we consider the OS uptime for the Docker version to symbolize the physical server, the reset of total traffic appears inconsistent. On the contrary, if the OS uptime for the container is the point of reference, the displayed OS uptime becomes inaccurate.

To address this inconsistency, I would like to propose two potential solutions. Firstly, incorporating total traffic information into a persistent volume could ensure that the Docker version behaves consistently with the standalone version. Alternatively, adjusting the OS uptime for the Docker version to reflect the uptime of the container could provide a more accurate representation.

I recognize that these proposals might introduce complexities, and I appreciate your expertise in assessing the viability of incorporating these adjustments. Should you determine that they do not align with the project's objectives, please feel free to close the issue. Despite this, I maintain the perspective that the current depiction of OS startup time, encompassing both container and host startup time, warrants reconsideration as I perceive it as a potential bug rather than a mere question.

Thank you for your time and consideration. I look forward to any further insights or decisions on this matter.

**Comment by Enkidu-6** [comment]
> The total upload/download traffic on the System Status page claims to represent the total network traffic since the system startup. However, it represents the total traffic since instance x-ui has been running

I don't see this behavior in my docker instance. I can restart the panel from the panel settings page, I can stop and restart Xray from the system status page and the traffic remains unchanged.

To top that off, I simply pruned my instance of docker and pulled a new one. The upload / download stats remained unchanged.

>To Reproduce

>  1.  Having a running instance of x-ui in docker.
>  2. Update the stack/container.

Updating the stack/container ! = restarting X-ui. It means building a new operating system from scratch. My guess is that you're not running your docker with network_mode: host. If you do that, The OS of your docker's host is the OS. If you don't, The OS of your docker instance is considered the OS and every time you rebuild you'll have a new OS.

If you run your container using docker compose and with network_mode: host none of these problems happen.

```

version: "3.9"

services:
  xui:
    image: alireza7/x-ui:latest
    container_name: x-ui
    hostname: YOUR_DOMAIN.TLD
    volumes:
      - $PWD/db/:/etc/x-ui/
      - $PWD/certs/:/etc/certs/
    environment:
      XRAY_VMESS_AEAD_FORCED: "false"
    tty: true
    network_mode: host
    restart: unless-stopped

```
Doing that, you won't see much difference between running X-ui as standalone or in docker. At least not until you try to run the x-ui menu from the command line. :-)

**Comment by alireza0** [comment]
@sgrtye Docker is an encapsulated OS and application inside has no access to the main OS.

@Enkidu-6 You are right. But `network_mode: host` is only about a kind of NAT in networking and never means that it is using main OS.

**Comment by Enkidu-6** [comment]
> @Enkidu-6 You are right. But `network_mode: host` is only about a kind of NAT in networking and never means that it is using main OS.

@alireza0 You're correct in that regard but NAT is not the only thing that's going on under the hood. It also means that your docker container is not isolated. Of course it has its own OS but for example in the host mode localhost is 127.0.0.1. In isolated mode localhost is the IP of your docker. Also a lot of commands run in the docker will refer to the host OS. For example typing ``` uptime ```  in the docker container in host mode will tell you the uptime of the Host and not the docker's OS.

**Comment by alireza0** [comment]
It is the result of this test in my MacOS:

Main system:
```bash
➜  ~ uptime
 9:25  up 2 days,  6:55, 3 users, load averages: 1.64 1.67 1.91
```

Nginx docker with --network=host:
```bash
root@docker-desktop:/# uptime
 08:25:34 up 10 min,  0 user,  load average: 0.10, 0.04, 0.00
```

Even load average is not same! DateTime is GMT in docker.

**Comment by Enkidu-6** [comment]
Here's mine on Almalinux

```
[root@3x2 ~]# uptime
 08:38:39 up 1 day,  1:51,  2 users,  load average: 0.00, 0.00, 0.00
[root@3x2 ~]# docker compose stop
[+] Stopping 1/0
 ✔ Container x-ui  Stopped
```
```
[root@3x2 ~]# uptime
 08:40:46 up 1 day,  1:53,  2 users,  load average: 0.00, 0.00, 0.00
[root@3x2 ~]# docker compose start
[+] Running 1/1
 ✔ Container x-ui  Started                                                                               0.2s
[root@3x2 ~]# docker exec -it x-ui sh
/app # uptime
 08:41:06 up 1 day,  1:54,  0 users,  load average: 0.00, 0.00, 0.00
```

**Comment by Enkidu-6** [comment]
And here's the result after completely pruning the container:

```
[root@3x2 ~]# docker compose stop
[+] Stopping 1/0
 ✔ Container x-ui  Stopped                                                                               0.1s
[root@3x2 ~]# docker system prune -a
WARNING! This will remove:
  - all stopped containers
  - all networks not used by at least one container
  - all images without at least one container associated to them
  - all build cache

Are you sure you want to continue? [y/N] y
Deleted Containers:
eded4f8aead8a8a757261b0840d842bfdcec5cd94e0c135729d934aed38ccecf

Deleted Images:
untagged: enkidu6/x-ui:latest
untagged: enkidu6/x-ui@sha256:2457ecef7c1c6bf31a0e13c6439498a98e42f26b37b504f4983f263bd66cd8fb
deleted: sha256:92f81b8e35a04a51369958f4f32fe9aebacab6c418e56c2a4903beaa2a93a45e
deleted: sha256:f3714e79d056aaab86575550c3725318607bc36344cdfa1f281b86e00c30d801
deleted: sha256:f2a749f8a2989130d283299b8b5034cfc10767aab78b0478e2a1e992cd8dbee6
deleted: sha256:76ef09126d24ba4000002e6c9d621b38327e452e27f36f84dc6df7f9c52f0e9c
deleted: sha256:5af4f8f59b764c64c6def53f52ada809fe38d528441d08d01c206dfb3fc3b691

Total reclaimed space: 131.9MB
[root@3x2 ~]# docker compose up -d
[+] Running 5/5
 ✔ xui 4 layers [⣿⣿⣿⣿]      0B/0B      Pulled                                                            4.7s
   ✔ 661ff4d9561e Pull complete                                                                          0.5s
   ✔ 2d245c07f937 Pull complete                                                                          0.4s
   ✔ e14799ba8ad3 Pull complete                                                                          0.5s
   ✔ 48a3f1e57dcf Pull complete                                                                          1.7s
[+] Running 1/1
 ✔ Container x-ui  Started                                                                               0.1s
[root@3x2 ~]# docker exec -it x-ui sh
/app # uptime
 08:47:49 up 1 day,  2:00,  0 users,  load average: 0.00, 0.00, 0.00
/app #
```

**Comment by Enkidu-6** [comment]
@alireza0 

P.S.
Just noticed you're using docker Desktop. 

From: https://docker-docs.uclv.cu/network/host/

"The host networking driver only works on Linux hosts, and is not supported on Docker Desktop for Mac, Docker Desktop for Windows, or Docker EE for Windows Server."

**Comment by alireza0** [comment]
I knew that it is not working properly in MacOS. I have mentioned it due to this fact.
But didn't know about other functions! It is not mentioned in docs!
Thats cool :)

**Comment by Enkidu-6** [comment]
Yes, On Docker Desktop, host mode is an imitation of the real thing and mimics a few things using NAT. On Linux on the other hand, the host mode means your docker container is no longer isolated. It's aware of its host and can interact with it at the OS level. The down side of it is that you may no longer have all the protections an isolated docker image provides.

In the isolated mode, getting the total upload / download for all network cards means just one network card, the docker's, which is why the upload / download resets when you build a new image. In the host mode, the container sees everything and it reports all the data for all network cards which is why it doesn't matter how many times you rebuild it. It talks to the host and gives you the accurate data.

Cheers.

**Comment by sgrtye** [comment]
@alireza0 @Enkidu-6,

Thank you both for your valuable information. Currently, I am running a stack in bridge network mode on Ubuntu. In this configuration, the container's uptime coincides with the host's uptime, while the container operates within a distinct network environment, resulting in a separate count for network traffic. The identified issue specifically arises when running x-ui on Docker in Linux with non-host network mode.

I have formulated a solution tailored to this particular setup. Instead of relying on 'uptime' to provide the OS uptime, we can utilize the Change time obtained from 'stat /proc' and calculate the difference between the Change time and the current time to derive the system's uptime.

The Change timestamp signifies the last time the file's inode has been changed. The ctime of /proc should remain constant and consistently represent the container's start time, accurately reflecting when the host is restarted. This adjustment aims to align the behaviors of Linux containers (running in non-host network mode) with non-Linux containers.

To mitigate potential risks, I propose introducing an additional environment variable to define the method for calculating OS uptime. Implementing this solution ensures compatibility with existing functionality and proves particularly beneficial for users utilizing Docker in non-host network mode on Linux.
